File: avl/avl.cpp
#include <iostream>
#include <fstream>
#include <chrono>

using namespace std;

// AVL Tree Node
struct Node
{
    int key;
    Node *left;
    Node *right;
    int height;

    Node(int k) : key(k), left(nullptr), right(nullptr), height(1) {}
};

// Function to get the height of a node
int getHeight(Node *node)
{
    if (node == nullptr)
        return 0;
    return node->height;
}

// Function to get the balance factor of a node
int getBalanceFactor(Node *node)
{
    if (node == nullptr)
        return 0;
    return getHeight(node->left) - getHeight(node->right);
}

// Function to update the height of a node
void updateHeight(Node *node)
{
    if (node == nullptr)
        return;
    node->height = max(getHeight(node->left), getHeight(node->right)) + 1;
}

// Function to perform a right rotation
Node *rightRotate(Node *y)
{
    Node *x = y->left;
    Node *T2 = x->right;

    x->right = y;
    y->left = T2;

    updateHeight(y);
    updateHeight(x);

    return x;
}

// Function to perform a left rotation
Node *leftRotate(Node *x)
{
    Node *y = x->right;
    Node *T2 = y->left;

    y->left = x;
    x->right = T2;

    updateHeight(x);
    updateHeight(y);

    return y;
}

// Function to balance the AVL tree
Node *balance(Node *node)
{
    if (node == nullptr)
        return nullptr;

    updateHeight(node);
    int balanceFactor = getBalanceFactor(node);

    // Left-Left case
    if (balanceFactor > 1 && getBalanceFactor(node->left) >= 0)
        return rightRotate(node);

    // Right-Right case
    if (balanceFactor < -1 && getBalanceFactor(node->right) <= 0)
        return leftRotate(node);

    // Left-Right case
    if (balanceFactor > 1 && getBalanceFactor(node->left) < 0)
    {
        node->left = leftRotate(node->left);
        return rightRotate(node);
    }

    // Right-Left case
    if (balanceFactor < -1 && getBalanceFactor(node->right) > 0)
    {
        node->right = rightRotate(node->right);
        return leftRotate(node);
    }

    return node;
}

// Function to insert a key into the AVL tree
Node *insert(Node *root, int key)
{
    if (root == nullptr)
        return new Node(key);

    if (key < root->key)
        root->left = insert(root->left, key);
    else if (key > root->key)
        root->right = insert(root->right, key);
    else
        return root; // Duplicate keys not allowed

    return balance(root);
}

// Function to find the node with minimum key value
Node *findMin(Node *node)
{
    while (node->left != nullptr)
        node = node->left;
    return node;
}

Node *deleteNode(Node *root, int key)
{
    if (root == nullptr)
        return root;

    if (key < root->key)
        root->left = deleteNode(root->left, key);
    else if (key > root->key)
        root->right = deleteNode(root->right, key);
    else
    {
        if (root->left == nullptr || root->right == nullptr)
        {
            Node *temp = root->left ? root->left : root->right;
            delete root;
            root = temp;
        }
        else
        {
            Node *temp = findMin(root->right);
            root->key = temp->key;
            root->right = deleteNode(root->right, temp->key);
        }
    }

    if (root == nullptr)
        return root;

    return balance(root);
}

// Function to find a key in the AVL tree
bool find(Node *root, int key)
{
    if (root == nullptr)
        return false;
    if (key == root->key)
        return true;

    if (key < root->key)
        return find(root->left, key);
    else
        return find(root->right, key);
}

// Function to perform an in-order traversal of the AVL tree
void inOrderTraversal(Node *root)
{
    if (root == nullptr)
        return;

    inOrderTraversal(root->left);
    // outFile << root->key << " ";
    cout << root->key << " ";
    inOrderTraversal(root->right);
}
void preOrderTraversal(Node *root)
{
    if (root == nullptr)
        return;
    cout << root->key;
    if (root->left != nullptr || root->right != nullptr)
    {
        cout << "(";
        preOrderTraversal(root->left);

        if (root->right != nullptr)
        {
            cout << ",";
            preOrderTraversal(root->right);
            cout << ")";
        }
        else
        {
            cout << ",)";
        }
    }
}

int main()
{
    freopen("in.txt", "r", stdin);
    freopen("out_avl.txt", "w", stdout);
    // ifstream inFile("in.txt");
    // ofstream outFile("out_avl.txt");
    ofstream reportFile("report_avl.txt");

    string operation;
    int key;

    Node *root = nullptr;
    std::chrono::duration<double, std::milli> iTime = std::chrono::duration<double, std::milli>::zero();
    std::chrono::duration<double, std::milli> dTime = std::chrono::duration<double, std::milli>::zero();
    std::chrono::duration<double, std::milli> fTime = std::chrono::duration<double, std::milli>::zero();
    std::chrono::duration<double, std::milli> tTime = std::chrono::duration<double, std::milli>::zero();

    while (cin >> operation)
    {

        if (operation == "I")
        {
            // inFile >> key;
            cin >> key;
            auto start = chrono::high_resolution_clock::now();
            root = insert(root, key);
            auto stop = chrono::high_resolution_clock::now();
            iTime += (stop - start);
            preOrderTraversal(root);
            // outFile << endl;
            cout << endl;
        }
        else if (operation == "D")
        {
            // inFile >> key;
            cin >> key;
            auto start = chrono::high_resolution_clock::now();
            root = deleteNode(root, key);
            auto stop = chrono::high_resolution_clock::now();
            dTime += (stop - start);
            preOrderTraversal(root);
            // outFile << endl;
            cout << endl;
        }
        else if (operation == "F")
        {
            // inFile >> key;
            cin >> key;
            auto start = chrono::high_resolution_clock::now();
            if (find(root, key))
                // outFile << "found" << endl;
                cout << "found" << endl;
            else
                cout << "not found" << endl;
            auto stop = chrono::high_resolution_clock::now();
            fTime += (stop - start);
        }
        else if (operation == "T")
        {
            auto start = chrono::high_resolution_clock::now();
            inOrderTraversal(root);
            auto stop = chrono::high_resolution_clock::now();
            tTime += (stop - start);
            // outFile << endl;
            cout << endl;
        }

        //     auto stop = chrono::high_resolution_clock::now();
        //     auto duration = chrono::duration_cast<chrono::microseconds>(stop - start);
    }
    reportFile << "operation time(ms)" << endl;
    reportFile << "Insert: "
               << " " << iTime.count() << endl;
    reportFile << "Delete: "
               << " " << dTime.count() << endl;
    reportFile << "Find: "
               << " " << fTime.count() << endl;
    reportFile << "Traversal: "
               << " " << tTime.count() << endl;
    reportFile << "Total: "
               << " " << iTime.count() + dTime.count() + fTime.count() + tTime.count() << endl;

    reportFile.close();

    return 0;
}

File: avl/avl.py
class AVLNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.height = 1


class AVLTree:
    def __init__(self):
        self.root = None

    def height(self, node):
        if node is None:
            return 0
        return node.height

    def balance_factor(self, node):
        if node is None:
            return 0
        return self.height(node.left) - self.height(node.right)

    def update_height(self, node):
        if node is None:
            return 0
        node.height = max(self.height(node.left), self.height(node.right)) + 1

    def rotate_left(self, z):
        y = z.right
        T2 = y.left

        y.left = z
        z.right = T2

        self.update_height(z)
        self.update_height(y)

        return y

    def rotate_right(self, y):
        x = y.left
        T2 = x.right

        x.right = y
        y.left = T2

        self.update_height(y)
        self.update_height(x)

        return x

    def rebalance(self, node):
        if node is None:
            return node

        self.update_height(node)

        balance = self.balance_factor(node)

        # Left-Left case
        if balance > 1 and self.balance_factor(node.left) >= 0:
            return self.rotate_right(node)

        # Right-Right case
        if balance < -1 and self.balance_factor(node.right) <= 0:
            return self.rotate_left(node)

        # Left-Right case
        if balance > 1 and self.balance_factor(node.left) < 0:
            node.left = self.rotate_left(node.left)
            return self.rotate_right(node)

        # Right-Left case
        if balance < -1 and self.balance_factor(node.right) > 0:
            node.right = self.rotate_right(node.right)
            return self.rotate_left(node)

        return node

    def insert(self, key):
        self.root = self._insert(self.root, key)

    def _insert(self, node, key):
        if node is None:
            return AVLNode(key)

        if key < node.key:
            node.left = self._insert(node.left, key)
        elif key > node.key:
            node.right = self._insert(node.right, key)
        else:
            return node  # Duplicate keys not allowed

        return self.rebalance(node)

    def find(self, key):
        return self._find(self.root, key)

    def _find(self, node, key):
        if node is None:
            return False
        if node.key == key:
            return True
        elif key < node.key:
            return self._find(node.left, key)
        else:
            return self._find(node.right, key)

    def delete(self, key):
        self.root = self._delete(self.root, key)

    def _delete(self, node, key):
        if node is None:
            return node

        if key < node.key:
            node.left = self._delete(node.left, key)
        elif key > node.key:
            node.right = self._delete(node.right, key)
        else:
            # Node with only one child or no child
            if node.left is None:
                temp = node.right
                node = None
                return temp
            elif node.right is None:
                temp = node.left
                node = None
                return temp

            # Node with two children: Get the inorder successor (smallest in the right subtree)
            temp = self._get_min_value_node(node.right)
            node.key = temp.key
            node.right = self._delete(node.right, temp.key)

        return self.rebalance(node)

    def _get_min_value_node(self, node):
        while node.left is not None:
            node = node.left
        return node

    def traversal(self):
        result = []
        self._in_order_traversal(self.root, result)
        return result

    def _in_order_traversal(self, node, result):
        if node:
            self._in_order_traversal(node.left, result)
            result.append(str(node.key))
            self._in_order_traversal(node.right, result)


def perform_operations(input_file):
    avl_tree = AVLTree()
    output_avl = open("out_avlpy.txt", "w")
    report_avl = open("report_avlpy.txt", "w")

    with open(input_file, "r") as f:
        for line in f:
            operation, value = line.strip().split(" ")

            start_time = time.time()

            if operation == "I":
                avl_tree.insert(int(value))
                output_avl.write("(" + avl_tree.root.__str__() + ")\n")
            elif operation == "D":
                avl_tree.delete(int(value))
                output_avl.write("(" + avl_tree.root.__str__() + ")\n")
            elif operation == "F":
                if avl_tree.find(int(value)):
                    output_avl.write("found\n")
                else:
                    output_avl.write("not found\n")
            elif operation == "T":
                traversal_result = avl_tree.traversal()
                output_avl.write(" ".join(traversal_result) + "\n")

            end_time = time.time()
            elapsed_time = (end_time - start_time) * 1000
            report_avl.write(f"{operation} {elapsed_time:.8f}\n")

    output_avl.close()
    report_avl.close()


if __name__ == "__main__":
    import time

    perform_operations("in.txt")
